[WIP-issue-224-ProcessMessage-refactor] first leg of the ProcessMessa…

gubatron opened the pull request:
After reading about ["Control Messages" and "Data Messages"](https://bitcoin.org/en/developer-reference#data-messages) I decided to name the base class "NetworkMessage" (and not "NetworkCommand")

After this has been revised I'll need help figuring out where this code should be moved to, perhaps it should live in its own pair of networkmessages.h, networkmessages.cpp files and I would need help adding those into the build machinery, I'm still a C++ noob.

Along the way I read some troubling things, one [I reported already for topic of discussion](https://github.com/bitcoinclassic/bitcoinclassic/issues/306).

gubatron committed:
[issue-244-refactor-network-handling-code] first leg of the ProcessMessage refactor

After reading about "Control Messages" and "Data Messages" I decided to name the base class "NetworkMessage" (and not "NetworkCommand")

After this has been revised I'll need help figuring out where this code should be moved to, perhaps it should live in its own pair of networkmessages.h, networkmessages.cpp files and I would need help adding those into the build machinery, I'm still a C++ noob.

gubatron commented:
Note: I won't be able to code on this for at least 10 more days, so no hurry.
#LeavingToAirportAsIPushThis

zander commented on src/main.cpp:
```diff
@@ -4155,26 +4156,58 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
+public:
+    virtual ~NetworkMessage() {}
+
+    NetworkMessage(CNode * const pFrom,
```
I think one constructor with only the pFrom should be fine.

More constructors and you just create problems for yourself because all the inheriting classes should mirror that.

The chainParams is a global singleton, I think we can stop passing it along in all of Bitcoin Classic.

Maybe the easiest solution for the booleans is to have some setters and make the constructor just default them to off.

gubatron replied:
almost done with the functional refactor (couldn't help myself and code this)

just to be on the safe side, I'm receiving the `chainParams` parameter. Not that sure it's a global singleton as I see it being instantiated in many different places

![screen shot 2017-09-20 at 7 20 28 am](https://user-images.githubusercontent.com/163977/30645859-45568f74-9dd4-11e7-9a45-85dbc895d1fd.png)

although internally `chainparams.cpp` it returns the same pointer when the global `Params()` function is invoked.

![screen shot 2017-09-20 at 7 21 51 am](https://user-images.githubusercontent.com/163977/30645878-5bd774fc-9dd4-11e7-80ac-e04a82faf5ed.png)

So another option I guess, is to not put that parameter in the method interface and just invoke `Params()` internally, but I like having the flexibility of passing any `CCHainParams` pointer, specially for testing it might come in handy.

zander replied:
The idea of having that flexibility in testing is however turns out to not be used as the testing framework sets the proper Params in the start of each test already and experience shows we don't want to change it halfway through a single test.

See BasicTestingSetup constructor where this is done.

gubatron replied:
ok, makes sense to get rid of this param and use `Params()` like I've seen you've been doing in recent fixes. 

Making this move now to `messages/` is letting me see what a clusterfuck there is in main.h and main.cpp.

zander commented on src/main.cpp:
```diff
@@ -4334,66 +4366,85 @@ bool static ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vR
-                        hashKey = Hash(BEGIN(hashKey), END(hashKey));
-                        mapMix.insert(std::make_pair(hashKey, pnode));
+        BOOST_FOREACH(CAddress &addr, vAddr) {
+                        boost::this_thread::interruption_point();
```
Please avoid whitespace changes like these.

zander commented on src/main.cpp:
```diff
@@ -4769,22 +4862,22 @@ bool static ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vR
-                }
-            }
-        }
+                        CValidationState state;
```
ditto. Too much whitespace changed.

gubatron replied:
hmm, thought the project was using spaces for indents, I see the line above has tabs.

How do you have your IDE configured to handle indentation? tabs? tabs turn into <n?> whitespaces?

zander replied:
Ah, tabs, that explains that..

My IDE (QtCreator) has the default-setting it only fixes whitespace on lines I changed anyway.

Fixing tabs should should interpret those tabs as 4 spaces.

Naturally, when the final code moves to another file, you'll end up touching lots of lines. But since your auto-indent for boost_foreach doesn't work, please make it look good manually :)

zander commented on src/main.cpp:
```diff
@@ -5319,42 +5514,211 @@ bool static ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vR
+    const bool xthinEnabled = IsThinBlocksEnabled();
+
+    NetworkMessage *networkMessage = nullptr;
+    if (strCommand == NetMsgType::VERSION)
```
You moved curly braces onto the same line as the if in another part, I'm a bit surprised you don't do that here too :)

I prefer curly braces on the same line as the if() statement.

gubatron replied:
ok, me too :) (coming from java)

zander commented on src/main.cpp:
```diff
@@ -5319,42 +5514,211 @@ bool static ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vR
-
-    return true;
+    // FIX-ME: Perhaps a try/catch here according to Ticket 224 description by @zander
+    return networkMessage->handle(vRecv, nTimeReceived, strCommand);;
```
Minor nit; double semicolon here.

Major nit; networkMessage may be null here.

gubatron replied:
will fix all the nit stuff and continue

zander reviewed (commented):
Hi,

I'm very happy with this already. Its making great progress and doing the classes 'in-line' makes for a very nice and readable (minimal) diff.

About the location;
the main purpose of this exercise is to make things maintainable by not having such huge methods or huge files.
I think the best way to go forward is to create a directory `messages/` and create a series of files there where each class is in its own file of the same name.
For instance VersionMessage.h and VersionMessage.cpp

ps. Any specific reason you based the names on the term 'message' instead of the term 'command'? I don't have any strong opinions either way, but its good to know if you had one I missed.

gubatron commented:
Yes, in all p2p protocols I've ever worked, these are called messages. I
don't see them as commands as these are simply messages exchanged by peers.
If valid then receiving peers may perform actions. I see as a command
something a user tells a local machine.

I'll improve this further over the weekend, headed back to the US in the
next 48 hours. I will make it so that this is a bit more functional in
style, I'm not too happy with all the state variables kept in the message
classes, I'll give it a try now passing more parameters to the handling
function, and each implementation will use these parameters as needed or
not.

On Sep 18, 2017 14:44, "TomZ" <notifications@github.com> wrote:

> *@zander* commented on this pull request.
>
> Hi,
>
> I'm very happy with this already. Its making great progress and doing the
> classes 'in-line' makes for a very nice and readable (minimal) diff.
>
> About the location;
> the main purpose of this exercise is to make things maintainable by not
> having such huge methods or huge files.
> I think the best way to go forward is to create a directory messages/ and
> create a series of files there where each class is in its own file of the
> same name.
> For instance VersionMessage.h and VersionMessage.cpp
>
> ps. Any specific reason you based the names on the term 'message' instead
> of the term 'command'? I don't have any strong opinions either way, but its
> good to know if you had one I missed.
> ------------------------------
>
> In src/main.cpp
> <https://github.com/bitcoinclassic/bitcoinclassic/pull/307#discussion_r139394639>
> :
>
> > -bool static ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, int64_t nTimeReceived)
> -{
> -    const CChainParams& chainparams = Params();
> -    RandAddSeedPerfmon();
> -    const bool fReindex = Blocks::DB::instance()->isReindexing();
> -    logDebug(Log::Net) << "received:" << SanitizeString(strCommand) << "bytes:" << vRecv.size() << "peer:" << pfrom->id;
> -    if (mapArgs.count("-dropmessagestest") && GetRand(atoi(mapArgs["-dropmessagestest"])) == 0)
> -    {
> -        LogPrintf("dropmessagestest DROPPING RECV MESSAGE\n");
> -        return true;
> -    }
> +class NetworkMessage {
> +public:
> +    virtual ~NetworkMessage() {}
> +
> +    NetworkMessage(CNode * const pFrom,
>
> I think one constructor with only the pFrom should be fine.
>
> More constructors and you just create problems for yourself because all
> the inheriting classes should mirror that.
>
> The chainParams is a global singleton, I think we can stop passing it
> along in all of Bitcoin Classic.
>
> Maybe the easiest solution for the booleans is to have some setters and
> make the constructor just default them to off.
> ------------------------------
>
> In src/main.cpp
> <https://github.com/bitcoinclassic/bitcoinclassic/pull/307#discussion_r139395157>
> :
>
> > -                        hashSalt = GetRandHash();
> -                    uint64_t hashAddr = addr.GetHash();
> -                    uint256 hashRand = ArithToUint256(UintToArith256(hashSalt) ^ (hashAddr<<32) ^ ((GetTime()+hashAddr)/(24*60*60)));
> -                    hashRand = Hash(BEGIN(hashRand), END(hashRand));
> -                    std::multimap<uint256, CNode*> mapMix;
> -                    BOOST_FOREACH(CNode* pnode, vNodes)
> -                    {
> -                        if (pnode->nVersion < CADDR_TIME_VERSION)
> -                            continue;
> -                        unsigned int nPointer;
> -                        memcpy(&nPointer, &pnode, sizeof(nPointer));
> -                        uint256 hashKey = ArithToUint256(UintToArith256(hashRand) ^ nPointer);
> -                        hashKey = Hash(BEGIN(hashKey), END(hashKey));
> -                        mapMix.insert(std::make_pair(hashKey, pnode));
> +        BOOST_FOREACH(CAddress &addr, vAddr) {
> +                        boost::this_thread::interruption_point();
>
> Please avoid whitespace changes like these.
> ------------------------------
>
> In src/main.cpp
> <https://github.com/bitcoinclassic/bitcoinclassic/pull/307#discussion_r139395459>
> :
>
> > -            if (pindexLast != NULL && header.hashPrevBlock != pindexLast->GetBlockHash()) {
> -                Misbehaving(pfrom->GetId(), 20);
> -                logWarning(Log::Net) << "p2p HEADERS command received non-continues headers sequence from" << pfrom->GetId();
> -                return false;
> -            }
> -            if (!AcceptBlockHeader(header, state, chainparams, &pindexLast)) {
> -                int nDoS;
> -                if (state.IsInvalid(nDoS)) {
> -                    if (nDoS > 0)
> -                        Misbehaving(pfrom->GetId(), nDoS);
> -                    logWarning(Log::Net) << "p2p HEADERS command received invalid header from" << pfrom->GetId();
> -                    return false;
> -                }
> -            }
> -        }
> +                        CValidationState state;
>
> ditto. Too much whitespace changed.
> ------------------------------
>
> In src/main.cpp
> <https://github.com/bitcoinclassic/bitcoinclassic/pull/307#discussion_r139396569>
> :
>
> >      }
>
> +    const bool xthinEnabled = IsThinBlocksEnabled();
> +
> +    NetworkMessage *networkMessage = nullptr;
> +    if (strCommand == NetMsgType::VERSION)
>
> You moved curly braces onto the same line as the if in another part, I'm a
> bit surprised you don't do that here too :)
>
> I prefer curly braces on the same line as the if() statement.
> ------------------------------
>
> In src/main.cpp
> <https://github.com/bitcoinclassic/bitcoinclassic/pull/307#discussion_r139397082>
> :
>
> >      else
>      {
>          // Ignore unknown commands for extensibility
>          logDebug(Log::Net) << "Unknown command" << SanitizeString(strCommand) << "from peer:" << pfrom->id;
>      }
>
> -
> -
> -    return true;
> +    // FIX-ME: Perhaps a try/catch here according to Ticket 224 description by @zander
> +    return networkMessage->handle(vRecv, nTimeReceived, strCommand);;
>
> Minor nit; double semicolon here.
>
> Major nit; networkMessage may be null here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoinclassic/bitcoinclassic/pull/307#pullrequestreview-63319475>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAicu5saejY-AYTobjx2hvX8jkgAXVks5sjleZgaJpZM4PQUcN>
> .
>

gubatron committed:
[issue-244-refactor-network-handling-code] whitespaces, indents, same-line if brackets

gubatron committed:
[issue-244-refactor-network-handling-code] NetworkMessage->handle() functional interface refactor, simpler, immutable

As of now it isn't necessary to keep any member variables in these single method classes. KISS.

gubatron commented:
Next up I'll do this:

>I think the best way to go forward is to create a directory messages/ and create a series of files there where >each class is in its own file of the same name.
>For instance VersionMessage.h and VersionMessage.cpp

gubatron commented:
perhaps inside `networkManager/messages/` ?

zander commented:
the networkManager dir is for the next generation network stack. This is all current generation.

It may be more confusing to put it in there as the NetworkManger is only used by the AdminServer right now.

zander commented:
The continues integration system reports that the QA test (the python stuff) fails the wallet.py test in this branch.

```
Running testscript wallet.py ...
[10:28:19]Initializing test directory /home/builder/buildAgent/temp/buildTmp/test5PINJJ
[10:28:19]start_node: bitcoind started, calling bitcoin-cli -rpcwait getblockcount
[10:28:19]['/home/builder/buildAgent/work/6ff1207d8200ef46/src/bitcoind', '-datadir=/home/builder/buildAgent/temp/buildTmp/test5PINJJ/node0', '-server', '-keypool=1', '-discover=0', '-rest', '-blockprioritysize=50000']
[10:28:20]start_node: calling bitcoin-cli -rpcwait getblockcount returned
[10:28:20]start_node: bitcoind started, calling bitcoin-cli -rpcwait getblockcount
[10:28:20]['/home/builder/buildAgent/work/6ff1207d8200ef46/src/bitcoind', '-datadir=/home/builder/buildAgent/temp/buildTmp/test5PINJJ/node1', '-server', '-keypool=1', '-discover=0', '-rest', '-blockprioritysize=50000']
[10:28:21]start_node: calling bitcoin-cli -rpcwait getblockcount returned
[10:28:21]start_node: bitcoind started, calling bitcoin-cli -rpcwait getblockcount
[10:28:21]['/home/builder/buildAgent/work/6ff1207d8200ef46/src/bitcoind', '-datadir=/home/builder/buildAgent/temp/buildTmp/test5PINJJ/node2', '-server', '-keypool=1', '-discover=0', '-rest', '-blockprioritysize=50000']
[10:28:22]start_node: calling bitcoin-cli -rpcwait getblockcount returned
[10:28:23]Mining blocks...
[10:28:27]Unexpected exception caught during testing: [Errno 111] Connection refused
[10:28:27]  File "/home/builder/buildAgent/work/6ff1207d8200ef46/qa/rpc-tests/test_framework/test_framework.py", line 135, in main
[10:28:27]    self.run_test()
[10:28:27]  File "/home/builder/buildAgent/work/6ff1207d8200ef46/qa/rpc-tests/wallet.py", line 76, in run_test
```

gubatron commented:
noted the wallet.py test issue.
I'll look into it once everything is finished inside `messages/`

zander commented:
Did you want any help with this?
I'll be focusing on the EDA issue for a little bit longer, but will try to find the reason for the wallet.py issue after that. Hopefully this week.

gubatron commented:
hi, no help needed, I just haven't been able to find time, last weekend I had a company get-together and couldn't even touch it.

On my local copy I've started moving everything already to their own files and I'm also cleaning up main as there are some functions there that pertain only to message management, but I want to be careful and do this with calm, so I'll need to get a good 2-3 hours of peace to do this.

What I need help with is on how to run that test with wallet.py `Running testscript wallet.py ...`, what command should I issue, it is a make variant, or is there a test script I can simply run (just for that particular test)

zander commented:
> What I need help with is on how to run that test with wallet.py

```
export PYTHON_DEBUG=1
./qa/pull-tester/rpc-tests.py wallet.py
```

And run `./qa/pull-tester/rpc-tests.py -h` as well for some flags. I guess adding `--nocleanup` will be useful.

So, these are the black-box testing ones. Which is tricky and fickle to work with. It always takes more time than expected for me ;)

gubatron committed:
[issue-244-refactor-network-handling-code] removed extra spaces from macro

gubatron committed:
[issue-244-refactor-network-handling-code] include optimization in checkpoints.cpp

gubatron commented:
big commit coming later today, will need help integrating all this to the Makefile machinery
![screen shot 2017-10-15 at 4 26 00 pm](https://user-images.githubusercontent.com/163977/31589871-a0c80a36-b1c5-11e7-95ba-89bc6e058581.png)

gubatron committed:
[issue-244-refactor-network-handling-code] moved all classes to messages/ folder, cleanup

zander committed:
[issue-244-refactor-network-handling-code] Add new files to automake-hell

zander commented on src/main.h:
```diff
@@ -550,4 +561,167 @@ static const unsigned int REJECT_ALREADY_KNOWN = 0x101;
 /** Transaction conflicts with a transaction already known */
 static const unsigned int REJECT_CONFLICT = 0x102;
 
+namespace {
```
defining an anonymous namespace in a header file can't work as the namespace will be local to each compile unit. Please give it a name or remove the namespace.

I want to say that its fine to leave this long list of stuff here, moving it elsewhere will cause me merge issues in another project. :)

gubatron replied:
thanks, I'm still trying to understand why in some places they have anonymous namespace and in some places things are declared globally, in some cases with `static` keyword

gubatron replied:
done

zander replied:
The problem that is being solves is one of symbol visibility.
Every method is a symbol and you want to avoid duplicates in the linked object (library/application).

The solution is to mark methods are local to the CPP file, which makes them compile-unit local and their symbols will not be seen by other compile-units.

Marking a method in a C++ file as 'static' does this trick.
The same result can be had by putting the methods in an anonymous namespace.
It should be noted that some compilers have bigger problems than others with polluting the global namespace, I believe gcc is pretty good overall. LLVM is even better. In general you won't really find it easy to hit the problem that is being solved.

Personally I prefer putting methods in an anonymous namespace to make clear they are file-local. Preferably all at the top of the file or so, or any other place where they are grouped.

In general; we use namespaces to avoid polluting (dumping everything into) the global namespace. This to avoid name-conflicts.

gubatron replied:
Thank you for this awesome explanation

zander commented on src/messages/NetworkMessage.h:
```diff
@@ -0,0 +1,312 @@
+
+static bool CheckIndexAgainstCheckpoint(const CBlockIndex* pindexPrev, CValidationState& state, const CChainParams& chainparams, const uint256& hash);
+
+namespace {
```
an anonymous namespace doesn't do what you want it to do in a header file.

Maybe something like
```
namespace Network {
  class Message {...};
 void  ProcessGetData(...);
}
```

Look at AdminServer.h for an example of namespaces if you want.

I think that most of the code from this file should go into a CPP file because anything implemented in a header file will implicitly be marked as 'inline', and that isn't useful for large methods.

As you move stuff into cpp files and that code is in a namespace, please consider reading
`/doc/developers-nodes.md`
which has a coding-style guideline.

gubatron replied:
👍

gubatron replied:
Quick question:

In AdminServer.cpp, they prefix every method with "Admin::" which is the namespace name.

Is it ok in my case to simply name the namespace, as all my .cpp's are already enclosed in

`namespace {`

so I'd do:

`namespace Network {`

instead of adding:

`Network::` as a prefix for every method and removing the existing `namespace {` at the top of the .cpps

zander commented on src/messages/NetworkMessage.h:
```diff
@@ -0,0 +1,312 @@
+        friend class CAddrMan;
+    };
+
+    // Used in InvMesssage, TxMessage and XThinBlockMessage
```
Does that mean this method could be protected?

gubatron replied:
started to do that to keep track of where these methods were actually used.

yes, so I was thinking maybe of doing something like that, making it protected, but then again it applies to only 3 out of the rest, I wouldn't want to put those methods on the parent NetworkMessage class.

zander commented on src/messages/HeadersMessage.cpp:
```diff
@@ -0,0 +1,145 @@
+                    }
+
+        if (pindexLast)
+            UpdateBlockAvailability(pfrom->GetId(), pindexLast->GetBlockHash());
```
You forgot to add a single line to the main.h making this main.cpp method available to the entire application.

gubatron replied:
ha, I had it and took it out because I was getting an unused warning

zander commented on src/messages/HeadersMessage.cpp:
```diff
@@ -0,0 +1,145 @@
+        const CChainParams chainparams = Params();
+        CBlockIndex *pindexLast = NULL;
+        BOOST_FOREACH(const CBlockHeader &header, headers) {
+                        CValidationState state;
```
Please be careful about your whitespace, we had this once before ;)

We use a really simple 4-spaces indent and never more than one indent for a block.

gubatron replied:
will have to revise again my autoformatting, will fix.

gubatron replied:
done.
Not sure why CLion might be doing this when I apply auto-format, everything has been set for 4 spaces, still looking into it.

zander commented on src/messages/HeadersMessage.h:
```diff
@@ -0,0 +1,42 @@
+namespace {
+    class HeadersMessage : public NetworkMessage {
+    public:
+        ~HeadersMessage() {}
```
no need to write an empty destructor, the compiler will write that for you if you omit it. The only reason to write such a thing is if you want to mark it virtual, which is only relevant in a baseclass.

zander commented on src/messages/InvMessage.cpp:
```diff
@@ -0,0 +1,144 @@
+                                    ss << inv2;
+                                    ss << filterMemPool;
+                                    pfrom->PushMessage(NetMsgType::GET_XTHIN, ss);
+                                    MarkBlockAsInFlight(pfrom->GetId(), inv.hash, chainparams.GetConsensus());
```
You forgot to declare the MarkBlockAsInFlight in the main.h

Which itself requires you move the namespace {} markers it is caught in right now so it can be used outside of the compile-unit.

zander reviewed (commented):
Great progress, I added the Makefile magic to your PR, and noted some reasons why it fails to compile right now.

gubatron committed:
[issue-244-refactor-network-handling-code] removed anonymous namespace added to main.h

gubatron committed:
[issue-244-refactor-network-handling-code] fixed indent on all BOOST_XXX blocks

(Not sure why CLion is treating those blocks differently when applying auto-format, all indents set to 4 spaces)

gubatron committed:
[issue-244-refactor-network-handling-code] all message classes now in Network:: namespace

Added missing license header
Next up will move implementations sitting inside Network.h to a new Network.cpp file and fix missing forward declarations

gubatron committed:
[issue-244-refactor-network-handling-code] new NetworkMessage.cpp

 - Moved all the inlined functions of the space into NetworkMessage.cpp
 - Added NetworkMessage.cpp to Makefile.am
 - Missing functions popped into main.cpp and were fixed by including the
   Network:: prefix on each call. I think it's a lot cooler now. Thanks @zander.

 I still want to see what else can be brought from main.cpp into NetworkMessage .h/.cpp that doesn't need to be there.

Still getting one build error.
```
src/script/bitcoinconsensus.cpp:88:114: error: no matching constructor for initialization of 'TransactionSignatureChecker'
        return VerifyScript(tx.vin[nIn].scriptSig, CScript(scriptPubKey, scriptPubKey + scriptPubKeyLen), flags, TransactionSignatureChecker(&tx, nIn), NULL);
```

gubatron committed:
[issue-244-refactor-network-handling-code] more cleanup

- moved CanDirectFetch and UpdateBlockAvailability into Network:: namespace (NetworkMessage .h/.cpp)
- removed redundant empty destructors on NetworkMessage subclasses
- Network:: methods laid out alphabetically in both .h and .cpp files
- optimized includes in main.cpp

gubatron committed:
[issue-244-refactor-network-handling-code] SendHeadersMessage boolean simplification

gubatron committed:
[issue-244-refactor-network-handling-code] MarkBlockAsInFlight and MarkBlockAsReceived moved to Network:: namespace

gubatron committed:
[issue-244-refactor-network-handling-code] moved more functions from main.cpp to Network::

CBlockIndex *FindForkInGlobalIndex(const CChain &chain, const CBlockLocator &locator);
void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vector<CBlockIndex*>& vBlocks, NodeId& nodeStaller);
bool PeerHasHeader(CNodeState *state, CBlockIndex *pindex);

gubatron committed:
[issue-244-refactor-network-handling-code] NULL -> nullptr. BOOST_FOREACH -> C++11 range for loop

Only done in NetworkMessage

gubatron committed:
[issue-244-refactor-network-handling-code] moved Misbehaving function from main.cpp to Network:: namespace

zander commented on src/main.h:
```diff
@@ -511,9 +520,6 @@ class CVerifyDB {
 };
 
-/** Find the last common block between the parameter chain and a locator. */
-CBlockIndex* FindForkInGlobalIndex(const CChain& chain, const CBlockLocator& locator);
```
Please don't move validation related methods to the network layer.

This one specifically is also used in init.cpp

gubatron replied:
woops, will move back

gubatron replied:
btw, I got curious to see in what state bitcoin-core was in this respect and they got rid of main.cpp completely. they moved most of this stuff to [`net_processing.cpp`](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp), for some reason, perhaps because it was 2am, I thought I had seen that method there.

What do you think about moving ProcessMessages and SendMessages and their utility methods from main.cpp into here (Network::)?

zander replied:
In another branch I have a refactor for the validation code, which is a super sensitive, and the result would be another couple thousand of lines of code moved out of main.cpp, so I agree with your goal.

I have not looked at the impact of moving those methods. So without knowing the full extend I can say that in general I agree.
I would advice to take it slow. Make things work and compile and all the tests pass between movings of code. Makes things easier in the end.

zander commented on src/main.h:
```diff
@@ -254,10 +254,19 @@ void UnlinkPrunedFiles(std::set<int>& setFilesToPrune);
+
+CBlockIndex* AddToBlockIndex(const CBlockHeader& block);
+
+enum FlushStateMode {
```
If you move the enum, you probably also want to move to the header the signature of its overload;
`bool static FlushStateToDisk(CValidationState &state, FlushStateMode mode)`

gubatron replied:
>If you move methods like Misbehaving from the global namespace to the Network namespace, its Ok to make the method start with a lowercase character and be consistent with the coding styleguide.

Love that, coming from java-land I hate uppercased methods :)

zander reviewed (commented):
Looking much better, great stuff.

Please also pay attention to compiler warnings, I see some that look like they could be bugs.

If you move methods like Misbehaving from the global namespace to the Network namespace, its Ok to make the method start with a lowercase character and be consistent with the coding styleguide.
Also notice that there are unit tests that call this method, but I guess you'll get to that as you make things compile ;)

gubatron committed:
[issue-244-refactor-network-handling-code] moved FindForkInGlobalIndex back to main

Validations methods should not be in the network layer.

I had originally moved this method here because it's overwhelmingly used on the network layer and only once in init.cpp, my bad.

gubatron committed:
[issue-244-refactor-network-handling-code] moved back FlushStateMode to main.cpp

not used elsewhere

gubatron committed:
[issue-244-refactor-network-handling-code] Network:: namespace methods to lower case

Also on VersionMessage.cpp I've put the code that was inside UpdatePreferredDownload(CNode* node, CNodeState* state) in the only place it was being used. If used only in one place it makes no sense to have another symbol for it

gubatron committed:
[issue-244-refactor-network-handling-code] fixed broken compilation, long -> bool cast on acceptBlockHeader() case return

gubatron commented:
```
./qa/pull-tester/rpc-tests.py --nocleanup --tracerpc wallet.py
...
DEBUG:BitcoinRPC:-439-> listsinceblock ["584582873e59b4819a1718f78555e4c7aad6f7a58ed39c63930d90953dd14af9"]
DEBUG:BitcoinRPC:<-439- {"lastblock": "584582873e59b4819a1718f78555e4c7aad6f7a58ed39c63930d90953dd14af9", "transactions": []}
Stopping nodes
DEBUG:BitcoinRPC:-440-> stop []
DEBUG:BitcoinRPC:<-440- "Bitcoin server stopping"
DEBUG:BitcoinRPC:-441-> stop []
DEBUG:BitcoinRPC:<-441- "Bitcoin server stopping"
DEBUG:BitcoinRPC:-442-> stop []
DEBUG:BitcoinRPC:<-442- "Bitcoin server stopping"
Tests successful
Duration: 74 s

```

Tests are passing over here now.

I'd like to move `SendMessage` from main.cpp out into the `Network` namespace as the last thing on this branch before a full revision and merge.

zander commented:
Pushed the current state to teamcity for building on all platforms and testing.

zander commented:
Linking seems to fail on linux;
> libbitcoin_server.a(libbitcoin_server_a-main.o):(.bss+0x480): multiple definition of `setBlockIndexCandidates'
> libbitcoin_server.a(libbitcoin_server_a-init.o):(.bss+0x280): first defined here

and a long list of other methods as well. 

> libbitcoin_server.a(libbitcoin_server_a-txorphancache.o):(.bss+0x20): multiple definition of `setDirtyBlockIndex'
> libbitcoin_server.a(libbitcoin_server_a-init.o):(.bss+0x100): first defined here

Clang gives problems too;

```
Undefined symbols for architecture x86_64:
  "FlushStateToDisk(CValidationState&, FlushStateMode)", referenced from:
      Network::TxMessage::handle(CNode*, CDataStream&, long long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, bool, bool) in libbitcoin_server.a(libbitcoin_server_a-TxMessage.o)
  "Network::State(int)", referenced from:
      Network::findNextBlocksToDownload(int, unsigned int, std::__1::vector<CBlockIndex*, std::__1::allocator<CBlockIndex*> >&, int&) in libbitcoin_server.a(libbitcoin_server_a-NetworkMessage.o)
      Network::processBlockAvailability(int) in libbitcoin_server.a(libbitcoin_server_a-NetworkMessage.o)
      Network::markBlockAsInFlight(int, uint256 const&, Consensus::Params const&, CBlockIndex*) in libbitcoin_server.a(libbitcoin_server_a-NetworkMessage.o)
      Network::markBlockAsReceived(uint256 const&) in libbitcoin_server.a(libbitcoin_server_a-NetworkMessage.o)
      Network::misbehaving(int, int) in libbitcoin_server.a(libbitcoin_server_a-NetworkMessage.o)
      Network::updateBlockAvailability(int, uint256 const&) in libbitcoin_server.a(libbitcoin_server_a-NetworkMessage.o)
      Network::InvMessage::handle(CNode*, CDataStream&, long long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, bool, bool) in libbitcoin_server.a(libbitcoin_server_a-InvMessage.o)
      ...
  "CheckIndexAgainstCheckpoint(CBlockIndex const*, CValidationState&, CChainParams const&, uint256 const&)", referenced from:
      Network::acceptBlockHeader(CBlockHeader const&, CValidationState&, CChainParams const&, CBlockIndex**) in libbitcoin_server.a(libbitcoin_server_a-NetworkMessage.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bitcoind] Error 1
```

zander commented:
Windows ends up in a compile error;

```
messages/AddrMessage.cpp: In member function 'virtual bool Network::AddrMessage::handle(CNode*, CDataStream&, int64_t, std::string&, bool, bool)':
messages/AddrMessage.cpp:48:13: error: 'interruption_point' is not a member of 'boost::this_thread'
             boost::this_thread::interruption_point();
```

I'm guessing a missing include.

gubatron commented:
yes, still fixing more compile errors.
I mistakenly ran the tests on master :p, therefore I deleted my last post here.

gubatron commented:
getting those same linking errors, will fix tonight

gubatron commented:
Update: This is still on my TODO list, Currently swamped with FrostWire 2.0's release but almost done over there. As soon as I get a breather I'll finish the current compiling/linking issues.

OT: what was the decision with EDA algo? I just read on reddit a new algorithm was picked without much technical review/simulation, I know this is one big ticket on this project.

gubatron commented:
fixed my linking issues (finally found some time today Sunday)
```
1 warning generated.
  AR       libbitcoin_util.a
  AR       libbitcoin_wallet.a
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libbitcoin_util.a(libbitcoin_util_a-strnlen.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libbitcoin_util.a(libbitcoin_util_a-strnlen.o) has no symbols
  AR       libbitcoin_zmq.a
  CXXLD    bitcoin-cli
  CXXLD    bitcoin-tx
  CXXLD    bitcoind
  CXXLD    test/test_bitcoin
make[1]: Nothing to be done for `all-am'.
```

now that I understand exactly what was happening, I'll continue to clean a bit more, make sure the tests pass, and hopefully get this ready for a review and merge

gubatron committed:
[issue-244-refactor-network-handling-code] Fixed linking issues, work remains

As I moved things into the messages/ folder, some static functions that were only meant to be
used in main.cpp were not being found in the Network:: namespace at linking time.

Such functions were:

    bool static FlushStateToDisk(CValidationState &state, FlushStateMode mode);

    static bool CheckIndexAgainstCheckpoint(const CBlockIndex* pindexPrev, CValidationState&...

and
    CNodeState *State(NodeId pnode), which was declared in an anonymous namespace that was meant
    only for main.cpp

This commit is an intermediate step now that I understand well where these functions are being used.

For example, FlushStateToDisk is used mostly on main.cpp and I don't like what I did on this commit, not making it static. On my next commit, I'll make it back into a static function meant only for main.cpp and I'll refactor the NetworkMessage class to indicate wheter it needs a call to FlushStateToDisk after myMessage.handle() has been called, only TxMessage needs a call to
FlushStateToDisk(state, FLUSH_STATE_PERIODIC) outside main.cpp

gubatron commented on src/main.cpp:
```diff
@@ -90,19 +90,15 @@ std::set<int> setDirtyFileInfo;
-std::map<NodeId, CNodeState> mapNodeState;
-
-// Requires cs_main.
 CNodeState *State(NodeId pnode) {
```
this guy got moved out of that anon namespace

gubatron commented on src/main.cpp:
```diff
@@ -1743,7 +1739,7 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
  * or always and in all cases if we're in prune mode and are deleting files.
  */
-bool static FlushStateToDisk(CValidationState &state, FlushStateMode mode) {
+bool FlushStateToDisk(CValidationState &state, FlushStateMode mode) {
```
this `static` is coming back, we won't need this outside main.cpp

gubatron commented on src/main.cpp:
```diff
@@ -2588,7 +2584,7 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
 }
 
-static bool CheckIndexAgainstCheckpoint(const CBlockIndex* pindexPrev, CValidationState& state, const CChainParams& chainparams, const uint256& hash)
+bool CheckIndexAgainstCheckpoint(const CBlockIndex* pindexPrev, CValidationState& state, const CChainParams& chainparams, const uint256& hash)
```
this one is coming back to, but I'll have to bring back acceptBlockHeader() from Network::

gubatron committed:
[issue-244-refactor-network-handling-code] acceptBlockHeader moved back to main

CheckIndexAgainstCheckpoint kept static inside main.cpp only. Still not sure what to do about FlushStateToDisk being needed only at the end of TxMessage::handle

gubatron commented:
**Update**: I think I got it to work like I wanted after thinking about it, writing my issue helped me think about it. sorry for the emails you may have received.

@zander need your C++ super powers on how to go about `bool FlushStateToDisk(CValidationState &state, FlushStateMode mode) `

This function is only used by one of the messages, `TxMessage`, it's used right at the end of its handle method.

I'd like to keep this function as it were before, `static` in `main.cpp`. 

is there a way to have it defined `static` and somehow being able to access its symbol on my other `Network::TxMessage::handle()` space?

My little C++ knowledge now makes me remove the `static` keyword, declare the function header in `main.h`, but then this yields a ton of unused warnings everywhere `main.h` is included, wish I could somehow just import that "private" symbol there.

The other way I'm thinking about it I don't really like, and it would involve having a method
```cpp
bool NetworkMessage::needsFlushing(CValidationState &state, FlushStateMode &mode);
```
that for all messages, except `TxMessage` would return `false`, which isn't very useful and that alone tells me it's probably a bad idea.

in `main.cpp`at the end of `ProcessMessage` I'd do something like this:
```cpp
bool result = networkMessage->handle(pfrom, vRecv, nTimeReceived, strCommand, xthinEnabled, fReindex);

CValidationState state;
FlushStateMode mode;
if (networkMessage->needsFlushing(state, mode)) {
    FlushStateToDisk(state, mode); // keeping all calls to FlushStateToDisk local to main.cpp
}

return result;
```

the default implementation of `needsFlushing()` would always return `false`, except for `TxMessage`, which would have to keep a reference of the `CValidationState state;` it creates while handle() is running on a private member variable

```cpp
bool TxMessage::needsFlushing(CValidationState &state, FlushStateMode &mode) {
    state = this->state;
    mode = FLUSH_STATE_PERIODIC;
    return true;
}
```

gubatron committed:
[issue-244-refactor-network-handling-code] no more unused FlushStateToDisk warning

Since FlushStateToDisk is only used internally in main.cpp it didn't need to be declared in the .h
and since it was also being used only in TxMessage.cpp the function couldn't be declared as statically linked
so all I did was a forward-declaration inside TxMessage.cpp and the symbol was found in both compilation and linking time

gubatron commented:
I think I got it :), I guess I wanted to do something that wasn't possible, to make a statically linked function available elsewhere. I simply hid the declaration from main.h and the unused warnings were gone.

gubatron commented:
Now onto this:
```
Mining blocks...
Unexpected exception caught during testing: [Errno 61] Connection refused
  File "/Users/gubatron/workspace.frostwire/bitcoinclassic/qa/rpc-tests/test_framework/test_framework.py", line 135, in main
    self.run_test()
```

zander commented:
run it with --nocleanup and check the logs of the nodes in the tmp dir.

gubatron commented:
awesome tip, it sure helped:

```
tail node*/regtest/debug.log
==> node0/regtest/debug.log <==
    .559 StopNode()
    .564 [Net] Flushed 0 addresses to peers.dat  5ms
    .575 [Coindb] Committing 104 changed transactions (out of 104) to coin database...
    .593 [DB] CDBEnv::Flush: Flush(true)
    .593 [DB] CDBEnv::Flush: Flushing wallet.dat (refcount = 0)...
    .597 [DB] CDBEnv::Flush: wallet.dat checkpoint
    .603 [DB] CDBEnv::Flush: wallet.dat detach
    .607 [DB] CDBEnv::Flush: wallet.dat closed
    .607 [DB] CDBEnv::Flush: Flush(true) took              14ms
    .620 Shutdown: done

==> node1/regtest/debug.log <==
    .586 [Bench]   - Writing chainstate: 0.02ms [0.00s]
    .586 [fees] Blockpolicy recalculating dynamic cutoffs: 
    .586 [fees] Blockpolicy after updating estimates for 2 confirmed entries, new mempool map size 0 
    .586 [Bitcoin] new best=1cd07a88e5fbb0a6f20fa5f7d8d76865142640d9917b6c72c67df37908bd01b5 height=103  log2_work=7.70044  tx=106  date=2017-11-06 02:27:56  progress=1  cache=0.0MiB(104tx)
    .586 [Bench]   - Connect postprocess: 0.20ms [0.03s]
    .586 [Bench] - Connect block: 0.65ms [0.06s]
    .586 [Mempool] Checking mempool with 0 transactions and 0 inputs
    .586 [2006] Processed Block 1cd07a88e5fbb0a6f20fa5f7d8d76865142640d9917b6c72c67df37908bd01b5 in 0.00 seconds
    .586 [Net] received: "notfound" bytes: 73 peer: 2 
    .586 [Net] Unknown command "notfound" from peer: 2 

==> node2/regtest/debug.log <==
    .587 [Net] disconnecting peer 1 
    .587 [Net] socket closed 
    .587 [Net] disconnecting peer 2 
02:27:41 [Net] socket closed 
    .621 [Net] disconnecting peer 3 
    .621 [Net] socket closed 
    .621 [Net] disconnecting peer 4 
02:27:42 [DB] Flushing wallet.dat
    .510 [DB] Flushed wallet.dat 24ms
02:28:37 Adding fixed seed nodes as DNS doesn't seem to be available.
```

From there we see the weird:
`    .586 [Net] Unknown command "notfound" from peer: 2 `

Which is my output on `ProcessMessage` for an unknown command.

This made me realize I had a bug, in that event the `networkMessage` object was a `nullptr` :)

```
start_node: calling bitcoin-cli -rpcwait getblockcount returned
Stopping nodes
bitcoind(65512,0x7fffb0aff3c0) malloc: *** error for object 0x7f7fb32eb760: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
bitcoind(65517,0x7fffb0aff3c0) malloc: *** error for object 0x7f87bbced9d0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
bitcoind(65515,0x7fffb0aff3c0) malloc: *** error for object 0x7f8cc2300650: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Tests successful
Duration: 70 s
```

Thank you!

gubatron committed:
[issue-244-refactor-network-handling-code] do not handle unknown messages

This is what was causing the crash on testscript wallet.py, an unknown message was received, our networkMessage object was not allocated and then we tried to handle the message with a nullptr.

however multi_rpc.py isn't passing yet

gubatron commented:
So that `wallet.py` test passes, but now the one for `multi_rpc.py` is breaking, will investigate:

```
Running testscript multi_rpc.py ...
Initializing test directory /var/folders/lw/hr00_r6j0xjdz481gq373tbc0000gn/T/test3EYfOI
init /var/folders/lw/hr00_r6j0xjdz481gq373tbc0000gn/T/test3EYfOI/node0
init /var/folders/lw/hr00_r6j0xjdz481gq373tbc0000gn/T/test3EYfOI/node1
init /var/folders/lw/hr00_r6j0xjdz481gq373tbc0000gn/T/test3EYfOI/node2
init /var/folders/lw/hr00_r6j0xjdz481gq373tbc0000gn/T/test3EYfOI/node3
start_node: bitcoind started, calling bitcoin-cli -rpcwait getblockcount
['/Users/gubatron/workspace.frostwire/bitcoinclassic/src/bitcoind', '-datadir=/var/folders/lw/hr00_r6j0xjdz481gq373tbc0000gn/T/test3EYfOI/node0', '-server', '-keypool=1', '-discover=0', '-rest']
start_node: calling bitcoin-cli -rpcwait getblockcount returned
start_node: bitcoind started, calling bitcoin-cli -rpcwait getblockcount
['/Users/gubatron/workspace.frostwire/bitcoinclassic/src/bitcoind', '-datadir=/var/folders/lw/hr00_r6j0xjdz481gq373tbc0000gn/T/test3EYfOI/node1', '-server', '-keypool=1', '-discover=0', '-rest']
Error: Unable to start HTTP server. See debug log for details.
start_node: calling bitcoin-cli -rpcwait getblockcount returned
start_node: bitcoind started, calling bitcoin-cli -rpcwait getblockcount
['/Users/gubatron/workspace.frostwire/bitcoinclassic/src/bitcoind', '-datadir=/var/folders/lw/hr00_r6j0xjdz481gq373tbc0000gn/T/test3EYfOI/node2', '-server', '-keypool=1', '-discover=0', '-rest']
start_node: calling bitcoin-cli -rpcwait getblockcount returned
start_node: bitcoind started, calling bitcoin-cli -rpcwait getblockcount
['/Users/gubatron/workspace.frostwire/bitcoinclassic/src/bitcoind', '-datadir=/var/folders/lw/hr00_r6j0xjdz481gq373tbc0000gn/T/test3EYfOI/node3', '-server', '-keypool=1', '-discover=0', '-rest']
start_node: calling bitcoin-cli -rpcwait getblockcount returned
Assertion failed: Block sync failed
  File "/Users/gubatron/workspace.frostwire/bitcoinclassic/qa/rpc-tests/test_framework/test_framework.py", line 133, in main
    self.setup_network()
  File "/Users/gubatron/workspace.frostwire/bitcoinclassic/qa/rpc-tests/test_framework/test_framework.py", line 67, in setup_network
    self.sync_all()
  File "/Users/gubatron/workspace.frostwire/bitcoinclassic/qa/rpc-tests/test_framework/test_framework.py", line 85, in sync_all
    sync_blocks(self.nodes)
  File "/Users/gubatron/workspace.frostwire/bitcoinclassic/qa/rpc-tests/test_framework/util.py", line 96, in sync_blocks
    raise AssertionError("Block sync failed")
Stopping nodes
bitcoind(64790,0x7fffb0aff3c0) malloc: *** error for object 0x7fe264ad7760: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
bitcoind(65789,0x7fffb0aff3c0) malloc: *** error for object 0x7faf3d5414a0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
bitcoind(65793,0x7fffb0aff3c0) malloc: *** error for object 0x7fa3f4a0ba40: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Failed
Traceback (most recent call last):
  File "./qa/pull-tester/rpc-tests.py", line 264, in <module>
    runtests()
  File "./qa/pull-tester/rpc-tests.py", line 164, in runtests
    rpcTestDir + testScripts[i] + flags, shell=True)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '/Users/gubatron/workspace.frostwire/bitcoinclassic/qa/rpc-tests/multi_rpc.py --srcdir /Users/gubatron/workspace.frostwire/bitcoinclassic/src   --nocleanup' returned non-zero exit status 1
```

gubatron committed:
typo

gubatron commented:
I'm stuck, I've gone back to the main branch and back to double check this wasn't there, and it's def. a bug on my branch. Everytime I `Ctrl+C` the `bitcoind` process it crashes shutting down. I always get this error:

```
bitcoind(94789,0x7fffb0aff3c0) malloc: *** error for object 0x7f9b9b3f1fd0: pointer being freed was not allocated
	*** set a breakpoint in malloc_error_break to debug
```

I've tried finding what/where is `0x7fffb0aff3c0` with gdb with no luck, I guess I don't know enough on how to use gdb to set up a breakpoint there. Wonder how I can do what the error message suggests, setting up a break point in "malloc_error_break" or if that's very generic and useless.

not sure what I may have done that has to do with the shutdown sequence.

any tips to debug this are welcome.

zander commented:
Memory issues like this are best debugged by running in valgrind.

Unfortunately it requires quite a bit of patience as it emulates your CPU.

In case you didn't realize, the error shows you are trying to delete something that was already deleted. (double delete).

gubatron commented:
so that same error message appeared when I tried to do:
`networkMessage->handle()` on a `null` pointer in process message, I don't recall doing any `delete` or `free` calls.

I'll have to find the code where `bitcoind` handles `SIGINT` and see how I can put a breakpoint on it to catch the spot where the double free or NPE occur.

Will check out valgrind!

gubatron committed:
[issue-244-refactor-network-handling-code] logic fix in ProcessMessages, missing returns in message handlers
